Initial attempt at a unit test for the EnvelopeEncryptionFilter

SamBarker opened the pull request:
ProduceRequestDataCondition feels like something we might want to follow up on regardless of wether you feel these tests are useful for the filter in question.

SamBarker committed:
Initial attempt at a unit test for the EnvelopeEncryptionFilter

Signed-off-by: Sam Barker <sbarker@redhat.com>

rh-pre-commit.version: 2.0.1
rh-pre-commit.check-secrets: ENABLED

SamBarker commented on kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/EnvelopeEncryptionFilter.java:
```diff
@@ -94,6 +98,9 @@ private CompletionStage<ProduceRequestData> maybeEncodeProduce(ProduceRequestDat
                     }).toList();
                     return join(futures).thenApply(x -> request);
+                }).exceptionallyCompose(throwable -> {
+                   log.warn("failed to encrypt records", throwable);
```
I needed this because I messed up the mocking and the exceptions were being silently swallowed. 

I'm not convinced that logging at warn is the entirely correct response but I think we should think about the error handling while encrypting.

tombentley replied:
I'm fine with adding this here for now. But can we add a similar `exceptionallyCompose` on the fetch side too, at the same time. So at least the "error handling" is being done the same on both paths.

tombentley commented on kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/EnvelopeEncryptionFilter.java:
```diff
@@ -36,12 +37,15 @@
  */
 public class EnvelopeEncryptionFilter<K>
         implements ProduceRequestFilter, FetchResponseFilter {
+    private final Logger log = getLogger(EnvelopeEncryptionFilter.class);
```
Any reason for this not to be `static`?

SamBarker replied:
Nope, just the intellij template I used.

SamBarker replied:
fixed in a36ab57eff5ac75c05bb01ba0f8bee43571a6289

tombentley commented on kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/EnvelopeEncryptionFilterTest.java:
```diff
@@ -0,0 +1,226 @@
+        encryptionFilter.onProduceRequest(ProduceRequestData.HIGHEST_SUPPORTED_VERSION, new RequestHeaderData(), produceRequestData, context);
+
+        // Then
+        verify(keyManager, times(0)).encrypt(any(), any(), any());
```
Is there not a `never()` rather than using `times(0)`?

SamBarker replied:
Yeah, I have fixed that in 31956245e7343dea605087526481fdfd05c6c4d3

Though it doesn't change the readability much.

tombentley reviewed (commented):
@SamBarker this looks like a good start, thank you! Can you add something for the fetch path? It doesn't have to be more than a single test, but enough that the rest of us have something to copy?

SamBarker committed:
Unit test the fetch path

Signed-off-by: Sam Barker <sbarker@redhat.com>

rh-pre-commit.version: 2.0.1
rh-pre-commit.check-secrets: ENABLED

SamBarker committed:
Make logger static

Signed-off-by: Sam Barker <sbarker@redhat.com>

rh-pre-commit.version: 2.0.1
rh-pre-commit.check-secrets: ENABLED

SamBarker commented:
> @SamBarker this looks like a good start, thank you! Can you add something for the fetch path? It doesn't have to be more than a single test, but enough that the rest of us have something to copy?

I've added something, but its not very satisfying yet. In particular it really wants the conditions I've added in https://github.com/kroxylicious/kroxylicious/pull/756 but there is a lot of other cruft to make things work.

SamBarker committed:
Add error logging on the fetch path.

Signed-off-by: Sam Barker <sbarker@redhat.com>

rh-pre-commit.version: 2.0.1
rh-pre-commit.check-secrets: ENABLED

SamBarker committed:
fix message for decode path

Signed-off-by: Sam Barker <sbarker@redhat.com>

rh-pre-commit.version: 2.0.1
rh-pre-commit.check-secrets: ENABLED
